### PR TITLE
ENH: Add new method to generate starting params

### PR DESCRIPTION
**Pull request by bashtage**
Added ability to use the long AR method in fitting ARIMA start_params
Improved tests. Fixed case where longar_maxlag is greater than length of
series
Changed longar_start_params variable name to start_ar_lags

closes #2550

### CODE REVIEW

File: statsmodels/tsa/tests/test_arima.py [hunk 1]
```diff
@@ -2233,13 +2233,27 @@ def test_arima_fit_mutliple_calls():
     y = [-1214.360173, -1848.209905, -2100.918158, -3647.483678, -4711.186773]
     mod = ARIMA(y, (1, 0, 2))
     # Make multiple calls to fit
-    with warnings.catch_warnings(record=True) as w:
-        mod.fit(disp=0, start_params=[np.mean(y), .1, .1, .1])
+    mod.fit(disp=0, start_params=[np.mean(y), .1, .1, .1])
     assert_equal(mod.exog_names,  ['const', 'ar.L1.y', 'ma.L1.y', 'ma.L2.y'])
-    with warnings.catch_warnings(record=True) as w:
-        mod.fit(disp=0, start_params=[np.mean(y), .1, .1, .1])
```
Comment (josef-pkt) [inline statsmodels/tsa/tests/test_arima.py]: Does this not raise a warning anymore?
Do you have an idea what changed?

Comment (josef-pkt) [inline statsmodels/tsa/tests/test_arima.py]: maybe rebase error
https://github.com/bashtage/statsmodels/commit/dff5609ce7f0bbbdbbded267aa7cf11b5d6b5512#diff-d0f09cb33612798898f77abc2f5ec297

File: statsmodels/tsa/tests/test_arima.py [hunk 2]
```diff
@@ -2233,13 +2233,27 @@ def test_arima_fit_mutliple_calls():
...
     assert_equal(mod.exog_names,  ['const', 'ar.L1.y', 'ma.L1.y', 'ma.L2.y'])
 
+def test_long_ar_start_params():
+    np.random.seed(12345)
+    arparams = np.array([.75, -.25])
+    maparams = np.array([.65, .35])
+
+    nobs = 30
+
+    y = arma_generate_sample(arparams, maparams, nobs)
```
Comment (josef-pkt) [inline statsmodels/tsa/tests/test_arima.py]: `arma_generate_sample` takes the lag polynomial parameterization, i.e. leading ones, `ar(L) y = ma(L) u`
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tsa/arima_process.py#L98

File: statsmodels/tsa/tests/test_arima.py [hunk 3]
```diff
@@ -2233,13 +2233,27 @@ def test_arima_fit_mutliple_calls():
...
+
+    model = ARMA(y, order=(2, 2))
+
+    res = model.fit(method='css',start_ar_lags=10)
+    res = model.fit(method='css-mle',start_ar_lags=10)
+    res = model.fit(method='mle',start_ar_lags=10)
```
Comment (josef-pkt) [inline statsmodels/tsa/tests/test_arima.py]: I guess this needs `disp=0`. It currently shows the lbfgsb printout in the test run.
https://travis-ci.org/statsmodels/statsmodels/jobs/140853476

File: statsmodels/tsa/tests/test_arima.py (hunk 1 above)
Comment (josef-pkt) [inline statsmodels/tsa/tests/test_arima.py]: The warning doesn't show up in the test run. But I don't know if there is no warning anymore, or whether it is not set to `always` warn

**Comment by josef-pkt** [comment]
I think it's better if we merge this (after fixing some details). It doesn't fix all the problems with the underlying code, but at least gives the option to work around some of them.

@chelleych

File: statsmodels/tsa/arima_model.py [hunk 4]
```diff
@@ -858,13 +876,20 @@ def fit(self, start_params=None, trend='c', method="css-mle",
             If True, all output from solver will be available in
             the Results object's mle_retvals attribute.  Output is dependent
             on the solver.  See Notes for more information.
-        disp : bool, optional
+        disp : int, optional
```
Comment (bashtage) [inline statsmodels/tsa/arima_model.py]: This was raising a warning in PyCharm since it was incorrectly labeled as bool

**Comment by bashtage** [comment]
No warnings are issued. I have no idea why those were there.  In general one should always check the error is raised when catching warnings.

**Comment by bashtage** [comment]
Recent Pythons get an error

```
statsmodels.tsa.tests.test_arima.test_arima_fit_mutliple_calls ... /home/travis/miniconda2/envs/statsmodels-test/lib/python3.4/site-packages/statsmodels-0.8.0-py3.4-linux-x86_64.egg/statsmodels/base/model.py:452: RuntimeWarning: Inverting hessian failed, no bse or cov_params available
  warn(warndoc, RuntimeWarning)
ok
```

File: statsmodels/tsa/arima_model.py (hunk 4 above)
Comment (josef-pkt) [inline statsmodels/tsa/arima_model.py]: My guess is that this depends on the scipy optimizer, but lbfgsb has numeric options.

File: statsmodels/tsa/arima_model.py [hunk 5]
```diff
@@ -732,7 +747,10 @@ def predict(self, params, start=None, end=None, exog=None, dynamic=False):
                 exog = exog[:, None]
 
         if out_of_sample != 0 and self.k_exog > 0:
-            # we need the last k_ar exog for the lag-polynomial
+            exog = np.asarray(exog)
+            if self.k_exog == 1 and exog.ndim == 1:
+                exog = exog[:, None]
+                # we need the last k_ar exog for the lag-polynomial
```
Comment (josef-pkt) [inline statsmodels/tsa/arima_model.py]: this part here is redundant/duplicate. It is now included in the lines just above under `if exog is not None`

**Comment by josef-pkt** [comment]
Given that python 3.4 shows the warning, the `catch_warning` needs to stay in the unit tests.

**Comment by bashtage** [comment]
https://github.com/statsmodels/statsmodels/blob/d2de2f0d91e89ff84856d86c2b2d2aa02e6c0fe5/statsmodels/base/model.py#L446

Is not correct.  There are many models in the test suite that have indefinite Hessians. This code will only show problems is an eigenvalue is very lose to 0, but when there are both + and - eigenvalues inv  "work".

See

https://travis-ci.org/bashtage/statsmodels/jobs/142324299

for all of the tests that have been incorrectly using indefinite Hessians.

**Comment by bashtage** [comment]
To understand 

```
statsmodels.base.tests.test_generic_methods.TestGenericNegativeBinomial.test_ttest_tvalues ... 0.999913988078 -0.94096264686
``` 

The numbers printing are the maximum and minimum eigenvalue of `Hinv`

**Comment by josef-pkt** [comment]
I'm not sure I understand.

First rebasing, amending or squashing during review makes it very difficult to follow what's going on.
https://mail.python.org/pipermail/scikit-learn/2016-June/000121.html

About Hessian with negative eigenvalues: 
AFAICS, this is a separate, new issue that is not directly related to the adding the start_params option to ARIMA.
Given the number of test errors the raise on negative eigenvalue causes, it will take some time to look at the incorrect hessian cases. 
However, many test cases like `test_arima_fit_mutliple_calls` don't need or use a valid hessian because they just test some specific feature or run a regression test for a specific item. (ARMA(1, 2) with 5 observations)
I have seen hessian with negative eigenvalues in some cases, but didn't realize that we have them pretty widespread in the test suite. This definitely needs a new issue and some design decisions.

**Comment by bashtage** [comment]
> First rebasing, amending or squashing during review makes it very difficult to follow what's going on.

I don't like writing commit messages.

**Comment by bashtage** [comment]
Back to passing, but ignoring the big issue.

If an explicit eigenvalue check is put in when convputing the inverse Hessian, there are many examples that have indefinite Hessians.

```
                _, v = np.linalg.eig(-1 * self.hessian(xopt))
                if np.min(v) < 0.0:
                    print(np.max(v), np.min(v))
                    raise RuntimeError
```

**Comment by bashtage** [comment]
The bigger issue is that warning catchers should not be allowed unless the source of the warning is fully understood and the warning is fully reproducible across platforms.  

If the warning differs depending on other modules (e.g. NumPy) then there should be a compat fix to unify the behavior.

**Comment by josef-pkt** [comment]
roughly: It takes a few seconds to write a commit message, and half an hour for a reviewer to figure out what changed.

(I'm also amending commits and force pushing, but not during review. Which is an empty statement because I usually don't get reviews.)

**Comment by josef-pkt** [comment]
>The bigger issue is that warning catchers should not be allowed unless the source of the warning is fully understood and the warning is fully reproducible across platforms.
If the warning differs depending on other modules (e.g. NumPy) then there should be a compat fix to unify the behavior.

open new issues with what should be changed so it doesn't get lost in PR discussion.

Right now we have no guarantee about consistent behavior in corner cases across machines, operating system, linalg libraries or versions of dependencies. For example, scipy optimizers change all the time, sometimes it handles a case better, sometimes worse. Differences in BLAS/LAPACK libraries can have quite large effects in edge cases.
The only thing that is currently guaranteed by our test suite are the numerical results, checked with assert_allclose or similar where precision is adjusted to the common denominator across computers and versions.
(To the latter: there were several difficult optimization problems in the test suite that didn't converge and unit tests failed or errored on some platforms and fixing them took quite a bit of time, mainly to figure out what might be the problem on a machine that we don't have access to.)

**Comment by josef-pkt** [comment]
Thanks @chelleych @bashtage 

AFAICS, this should be good to backport to 0.8.
